onnxruntime/core/providers/cpu/nn/shrink.cc

```diff
@@ -0,0 +1,99 @@
+
+#include "core/providers/cpu/nn/shrink.h"
+
+#include "core/util/math.h"
+#include "core/util/math_cpuonly.h"
+
+namespace onnxruntime {
+ONNX_CPU_OPERATOR_KERNEL(
+    Shrink,
+    9,
+    KernelDefBuilder().TypeConstraint("T", {DataTypeImpl::GetTensorType<float>(),
+                                            DataTypeImpl::GetTensorType<double>(),
+                                            DataTypeImpl::GetTensorType<int64_t>(),
+                                            DataTypeImpl::GetTensorType<uint64_t>(),
+                                            DataTypeImpl::GetTensorType<int32_t>(),
+                                            DataTypeImpl::GetTensorType<uint32_t>(),
+                                            DataTypeImpl::GetTensorType<int16_t>(),
+                                            DataTypeImpl::GetTensorType<uint16_t>(),
+                                            DataTypeImpl::GetTensorType<int8_t>(),
+                                            DataTypeImpl::GetTensorType<uint8_t>(),
+                                            DataTypeImpl::GetTensorType<MLFloat16>(),
+                                            DataTypeImpl::GetTensorType<BFloat16>()}),
+    Shrink);
+
+namespace shrink_internal {
+template <class T>
+inline T ShrinkImpl(const T& val, float bias, float lambd) {
+  // The ONNX spec doesn't take numeric overflow and underflow into account
+  // Implementing the spec as is for now
+  if (val < -lambd) {
+    return T(val + bias);
+  } else if (val > lambd) {
+    return T(val - bias);
+  } else {
+    return T(0);
+  }
+}
+
+void ShrinkMLFloat16(const Tensor* input, Tensor* output, float bias, float lambd) {
+  auto span = gsl::make_span(input->Data<MLFloat16>(), input->Shape().Size());
+  auto output_data = output->template MutableData<MLFloat16>();
+  std::transform(span.cbegin(), span.cend(), output_data, [bias, lambd](const MLFloat16& val) {
+    float fl = math::halfToFloat(val.val);
+    return MLFloat16(math::floatToHalf(ShrinkImpl<float>(fl, bias, lambd)));
+  });
+}
+
+void ShrinkBFloat16(const Tensor* input, Tensor* output, float bias, float lambd) {
+  auto span = gsl::make_span(input->Data<BFloat16>(), input->Shape().Size());
+  auto output_data = output->template MutableData<BFloat16>();
+  std::transform(span.cbegin(), span.cend(), output_data, [bias, lambd](const BFloat16& val) {
+    float fl = val.ToFloat();
+    return BFloat16(ShrinkImpl<float>(fl, bias, lambd));
+  });
+}
+
+}  // namespace shrink_internal
+
+Status Shrink::Compute(OpKernelContext* p_op_kernel_context) const {
+  using namespace shrink_internal;
+
+  auto input = p_op_kernel_context->Input<Tensor>(0);
+  auto output = p_op_kernel_context->Output(0, input->Shape());
+
+  auto dtype = input->DataType();
+
+  if (dtype == DataTypeImpl::GetType<float>()) {
+    EigenMap<float>(*output) = EigenMap<float>(*input).unaryExpr([this](const float& val) { return ShrinkImpl<float>(val, bias_, lambd_); });
```

Comment:
Very optional suggestion: 

You could reduce the code here by using DispatchOnTensorType (utils.h), and having a generic Shrink method that calls EigenMap, with two specializations for BFloat16 and MLFloat16 to special case those.

Reply:
Cannot directly use `DispatchOnTensorType` as it supports `bool` internally, but went for something similar. Thanks for this tip.

Reply:
I don't think it's necessary to add DispatchOnNumericTensorType as the ONNX type constraint validation should have rejected any input that was a bool much earlier. You could temporarily hack a unit test to verify that is the case. 

If you still don't trust that, could we do something slightly different to reduce the duplication in utils.h and instead use DispatchOnTensorTypeWithReturn, change ShinkImpl to return Status, and also specialize ShrinkImpl on bool and return an error if that is hit?

Reply:
When I used `DispatchOnTensorType` originally, the build resulted in a 'warning treated as error' as when it tried to instantiate ShrinkCore<bool>, it resulted in warnings because the implementation contained '>' and '<' which is considered unsafe for bool types.

So, going with the second option of returning `Status` in ShrinkImpl and specializing ShrinkImpl for `bool` and return an error for that case.